### ISSUE TITLE
Warning CS1701

### ISSUE THREAD

**Issue by andead** [cake-contrib/Cake.ExtendedNuGet#35]
When the cake script is compiled, CS1701 warnings show in the output.

## Expected Behavior
There are no warnings in the output.

## Current Behavior
6 compiler warnings CS1701 regarding NuGet.Versioning and NuGet.Packaging.

## Steps to Reproduce (for bugs)

1. Install Cake tool 

```
dotnet tool install --global Cake.Tool
```

2. Create a simple `test.cake` script

```
#addin "Cake.ExtendedNuGet"

Task("Default")
    .Does(() => 
    {
        var packageList = NuGetList("Cake");
        foreach(var package in packageList)
        {
            Information("Found package {0}, version {1}", package.Name, package.Version);
        }
    })

RunTarget("Default");
```

3. Run Cake

```
dotnet cake test.cake
```

4. Notice the output

```
(2671,12): warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy
(2677,12): warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy
(2683,12): warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy
(2689,12): warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy
(2707,12): warning CS1701: Assuming assembly reference 'NuGet.Packaging, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Packaging, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Packaging', you may need to supply runtime policy
(2713,12): warning CS1701: Assuming assembly reference 'NuGet.Packaging, Version=5.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Packaging, Version=5.3.0.4, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Packaging', you may need to supply runtime policy
```

## Context
I was trying to use `IsNuGetPublished` method in my Cake script and minimize the number of warnings in the output.

## Your Environment
docker image: mcr.microsoft.com/dotnet/core/sdk:3.0
Cake 0.35.0

* Module version used: 2.1.1
* Operating System and PowerShell version: Windows 10 Enterprise, Powershell not used

**Comment by steverichey** [comment]
I am receiving an identical warning with Cake 0.38.4 on macOS 10.15.6 in zsh 5.8 (x86_64-apple-darwin19.3.0).

**Comment by augustoproiete** [comment]
This was resolved by #37 which bumps NuGet.* dependencies to 5.4.0 (to match the same versions that Cake 0.36.0 to 0.38.5 use) and will be available in the next release

**Comment by hotchkj** [comment]
Rather than open a new issue, I wanted to query - I'm seeing this same result now with later versions of Cake e.g. `1.1.0`. Is this a moving target where these packages will often be out of sync and/or is there intent to regularly update to match?

Is there something that needs requesting of Cake to make this easier? Could they publish an assembly that gives you the matching NuGet references for example? Just throwing out ideas. Thanks for any available assistance.

```
warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.4.0.3, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.8.1.7021, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy
```

**Comment by augustoproiete** [comment]
Hey @hotchkj!

The short answer is that this warning is to be expected. Cake v1.1.0 references a newer version of the NuGet packages (v5.8.1) whilst the latest version of the `Cake.ExtendedNuGet` addin (v4.0.2) references an older version of NuGet (v5.4.0).

In general **this warning can be safely ignored** because Cake uses a newer version that is still compatible with the version this addin needs. It would be different if it was the other way around, e.g. If this addin needed a higher version than what Cake uses.

We might be able to suppress this warning in Cake in a future version (needs investigation). [Devs in the .NET team started discussing removing the CS1701 and CS1702 warnings back in 2017](https://github.com/dotnet/roslyn/issues/19640) but didn't happen yet.

On every release of Cake, we update to the latest stable NuGet packages available at the time, so this warning will continue to be shown until we can suppress it, or until a new version of `Cake.ExtendedNuGet` is released using a matching version.

We reference the lowest possible version of the NuGet packages in `Cake.ExtendedNuGet` for broader compatibility with the different versions of Cake. By referencing NuGet v5.4.0 we ensure that this addin works on Cake v0.38.5, v1.0.0, and v1.1.0 - and likely any new version of Cake released until a v6.x of the NuGet packages is out, at which point we'll be forced to bump the NuGet packages here in this addin too.

Does that make sense?

I think the way forward would be to invest in:

- Suppressing the CS1701 warning in Cake for cases like these. At a minimum (assuming it's possible) update Cake to only display these specific warnings when running on Diagnostic mode, to at least remove the noise on regular runs

and/or

- Consider bringing all (or some) of the features of `Cake.ExtendedNuGet` into the Cake codebase and ship these features as part of Cake, and possibly deprecate this addin

---

It might be better to continue the discussion here if you're interested: https://github.com/cake-build/cake/discussions

**Comment by hotchkj** [comment]
Thanks for the timely reply @augustoproiete! That makes sense, and I agree the right approach is to discuss with Cake about getting the warnings suppressed or similar strategies. They aren't really adding anything other than noise.

**Comment by augustoproiete** [comment]
@hotchkj PR https://github.com/cake-build/cake/issues/3337 got merged and starting with Cake v1.2.0 (expected to be released in a few days) these warnings will be suppressed.

Thanks for bringing this up

**Comment by augustoproiete** [comment]
@hotchkj [Cake v1.2.0 has been released](https://cakebuild.net/blog/2021/08/cake-v1.2.0-released) and fixes the above
